### PR TITLE
Hedera Token Service Precompiled Contract for Hedera Smart   Contract Service

### PR DESCRIPTION
**Pull request by shemnon**
**Description**:

Describe the integration of Hedera Token Service (HTS) with the Hedera Smart
Contract Service (HSCS), allowing contracts to transfer, mint, burn, associate,
and dissociate tokens programmatically.

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: HIP/hip-0000-hts-precompile.md [hunk 1]
```diff
@@ -0,0 +1,299 @@
+- hip: <HIP number (this is determined by the HIP editor)>
```
Comment (sergmetelin) [inline HIP/hip-0000-hts-precompile.md:1]: Please change the formatting to `---` wrap around the header and remove `-` on each line (as shown in https://raw.githubusercontent.com/hashgraph/hedera-improvement-proposal/master/hip-0000-template.md)

File: HIP/hip-0000-hts-precompile.md [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+- title: Hyperledger Token Service Precompiled Contract for Hyperledger Smart
+  Contract Service
+- author: Danno Ferrin <danno.ferrin@gmail.com>
+- type: Standards Track
+- category: Core
+- status: Draft
```
Comment (sergmetelin) [inline HIP/hip-0000-hts-precompile.md:7]: Please change status to `Review` once ready for the editorial review

File: HIP/hip-0000-hts-precompile.md [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+- created: 2021-11-04
+- discussions-to: &lt;a URL pointing to the official discussion thread>
+- updated:
```
Comment (sergmetelin) [inline HIP/hip-0000-hts-precompile.md:10]: Please remove unused fields

File: HIP/hip-0000-hts-precompile.md [hunk 4]
```diff
@@ -0,0 +1,299 @@
+- hip: <HIP number (this is determined by the HIP editor)>
+- title: Hyperledger Token Service Precompiled Contract for Hyperledger Smart
+  Contract Service
+- author: Danno Ferrin <danno.ferrin@gmail.com>
+- type: Standards Track
+- category: Core
+- status: Draft
+- created: 2021-11-04
+- discussions-to: &lt;a URL pointing to the official discussion thread>
```
Comment (sergmetelin) [inline HIP/hip-0000-hts-precompile.md:9]: Please add a link to the discussion thread

File: HIP/hip-0000-hts-precompile.md (hunk 1 above)
Comment (sergmetelin) [inline HIP/hip-0000-hts-precompile.md:1]: This HIP will be assigned number `206`, please rename the file to `hip-206.md`

File: assets/hip-hts/solidity/HederaResponseCodes.sol [hunk 5]
```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (sergmetelin) [inline assets/hip-hts/solidity/HederaResponseCodes.sol:1]: Please rename the folder from `hip-hts` to `hip-206`

**Review by sergmetelin — CHANGES_REQUESTED** [review]

**Comment by steven-sheehy** [comment]
Isn't the PR title supposed to be `s/Hyperledger/Hedera/g`?

**Comment by shemnon** [comment]
Years of muscle memory.  Blockchain and Letter H has meant something else to me for 3 years.

File: HIP/hip-0000-hts-precompile.md (hunk 1 above)
Comment (shemnon) [reply HIP/hip-0000-hts-precompile.md]: done

File: HIP/hip-0000-hts-precompile.md (hunk 3 above)
Comment (shemnon) [reply HIP/hip-0000-hts-precompile.md]: done

File: HIP/hip-0000-hts-precompile.md (hunk 4 above)
Comment (shemnon) [reply HIP/hip-0000-hts-precompile.md]: done

File: HIP/hip-0000-hts-precompile.md (hunk 1 above)
Comment (shemnon) [reply HIP/hip-0000-hts-precompile.md]: done

File: assets/hip-hts/solidity/HederaResponseCodes.sol (hunk 5 above)
Comment (shemnon) [reply assets/hip-hts/solidity/HederaResponseCodes.sol]: done

File: HIP/hip-206.md [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+> - ~~[ ] Support tokenUpdate?~~ Not in the first pass
+> - ~~[ ] Support token pause/unpause?~~ Not in the first pass
+
+### Contract Key
+
+To support contracts controlling features like minting and burning the defined
+but up until now unused field `contractID` in the `Key` protobuf message. When a
+token has a contract key as the admin key or one of the admin keys then that key
+is considered validated if the `CALLER` of the relevant method is the contract
+identified in the call. as well as the sender of the contract itself. This is
```
Comment (Daniel-K-Ivanov) [inline HIP/hip-206.md:129]:
```suggestion
identified in the call. This is
```

File: HIP/hip-206.md [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+library is fully known and audited ahead of time.
+
+### Signature Verification
+
+For operations that require signature verification the signatures of the top
+level contract call are considered, in addition to the immediate caller of the
+operation. This may result in a contract call that may need to be signed by more
+than the initial caller of the smart contract.
+
+In order for a signature to be used for verification the entire public key needs
```
Comment (Daniel-K-Ivanov) [inline HIP/hip-206.md:155]:
```suggestion
In order for a signature to be used for verification, the entire public key needs
```

File: HIP/hip-206.md [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+to be stored in the `pubKeyPrefix` field of the `SignaturePair` message. Any
+partial prefixes provided in the signature map will be ignored and will not
+properly authorize the contract, even if the signature is valid. This is the
+current practice of the Hedera supplied SDKs.
+
+### ERC-20 calls directly to Token Accounts
+
+To enable HTS tokens to be more bradly used in DeFi applications the
```
Comment (Daniel-K-Ivanov) [inline HIP/hip-206.md:163]:
```suggestion
To enable HTS tokens to be more broadly used in DeFi applications the
```

**Review by Daniel-K-Ivanov — COMMENTED** [review]
🚀 Great Job! I've added my comments in the discussion URL.

**Review by sergmetelin — APPROVED** [review]
LGTM, added Last Call status

**Pull request merged**